server/src/main/java/org/elasticsearch/index/engine/ThreadPoolMergeScheduler.java

```diff
@@ -0,0 +1,482 @@
+            return null;
+        }
+        MergeTask mergeTask = queuedMergeTasks.poll();
+        if (mergeTask == null) {
+            // no more merges to execute
+            return null;
+        }
+        assert mergeTask.isRunning() == false;
+        currentlyRunningMergeTasks.add(mergeTask);
+        return mergeTask;
+    }
+
+    private void submitNewMergeTask(MergeSource mergeSource, MergePolicy.OneMerge merge, MergeTrigger mergeTrigger) {
+        MergeTask mergeTask = newMergeTask(mergeSource, merge, mergeTrigger);
+        enqueMergeTask(mergeTask);
+        maybeActivateThrottling();
+    }
+
+    private void mergeDone(MergeTask mergeTask) {
+        assert mergeTask.isRunning();
+        update(() -> this.currentlyRunningMergeTasks.remove(mergeTask));
+        maybeDeactivateThrottling();
+    }
+
+    private void maybeActivateThrottling() {
+        int numRunningMerges = currentlyRunningMergeTasks.size();
+        int numQueuedMerges = queuedMergeTasks.size();
+        int configuredMaxMergeCount = maxThreadCount;
+        // both currently running and enqueued count as "active" for throttling purposes
+        if (numRunningMerges + numQueuedMerges > configuredMaxMergeCount && shouldThrottleIncomingMerges.getAndSet(true) == false) {
+            activateThrottling(numRunningMerges, numQueuedMerges, configuredMaxMergeCount);
+        }
+    }
+
+    private void maybeDeactivateThrottling() {
+        int numRunningMerges = currentlyRunningMergeTasks.size();
+        int numQueuedMerges = queuedMergeTasks.size();
+        int configuredMaxMergeCount = maxMergeCount;
+        // both currently running and enqueued count as "active" for throttling purposes
+        if (numRunningMerges + numQueuedMerges <= configuredMaxMergeCount && shouldThrottleIncomingMerges.getAndSet(false)) {
+            deactivateThrottling(numRunningMerges, numQueuedMerges, configuredMaxMergeCount);
+        }
+    }
+
+    private MergeTask newMergeTask(MergeSource mergeSource, MergePolicy.OneMerge merge, MergeTrigger mergeTrigger) {
+        // forced merges, and merges triggered when closing shard, always run un-throttled
+        boolean isAutoThrottle = mergeTrigger != MergeTrigger.CLOSING && merge.getStoreMergeInfo().mergeMaxNumSegments() == -1;
+        return new MergeTask(
+            mergeSource,
+            merge,
+            isAutoThrottle,
+            "Lucene Merge Task #" + mergeTaskCount.incrementAndGet() + " for shard " + shardId
+        );
+    }
+
+    private void enqueMergeTask(MergeTask mergeTask) {
+        assert mergeTask.isRunning() == false;
+        update(() -> this.queuedMergeTasks.add(mergeTask));
+    }
+
+    private void update(Runnable updater) {
+        threadPoolMergeExecutor.updateMergeScheduler(this, (ignored) -> {
+            synchronized (this) {
+                updater.run();
+            }
+        });
+    }
```

Comment:
Any update to any merge scheduler first takes the mutex of the single `ThreadPoolMergeExecutor` instance, and then of the merge scheduler itself.